### PR TITLE
Chapter2 Solutions!!!

### PR DESCRIPTION
**Pull request by ynfle**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -348,8 +351,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l = take (len `div` 2) l
+    where
+        len = length l
```
Comment (vrom911) [inline src/Chapter2.hs:357]: Great usage of `where` construction! Makes code easily readable 💯

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,7 +628,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [Int] -> [Int]
+takeEven [] = []
+takeEven (x:_:xs) = x : takeEven xs
+takeEven [x] = x : []
```
Comment (vrom911) [inline src/Chapter2.hs:634]: Great pattern matching skills 💯 

Also, `[x]` is the same as `x : []` 🙂

File: src/Chapter2.hs [hunk 3]
```diff
@@ -727,7 +738,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\x -> replicate x x) l)
```
Comment (vrom911) [inline src/Chapter2.hs]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 4]
```diff
@@ -740,7 +751,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains n l = filter (elem n) l
```
Comment (vrom911) [inline src/Chapter2.hs]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 5]
```diff
@@ -780,13 +792,15 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
+divideTenBy = div 10
 
 -- TODO: type ;)
+listElementsLessThan :: Int -> [Int] -> [Int]
 listElementsLessThan x l = filter (< x) l
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Int] -> [Int] -> [Int] 
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:803]: Eta-reduction award 🏆

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +877,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Fantastic work with list! 🤩 
Really nice solutions!

File: src/Chapter2.hs (hunk 2 above)
Comment (ynfle) [reply src/Chapter2.hs:634]: Thanks. Is one of them preferred?

File: src/Chapter2.hs (hunk 3 above)
Comment (ynfle) [reply src/Chapter2.hs]: Oh wow, that's a really nice solution. I'll try it out.
Is there another way to do this other than using `concat`?

File: src/Chapter2.hs (hunk 6 above)
Comment (ynfle) [reply src/Chapter2.hs]: I will think about this. Thanks

File: src/Chapter2.hs (hunk 2 above)
Comment (vrom911) [reply src/Chapter2.hs:634]: Usually, `[x]` is used. Though it is desugared to `x : []` as you wrote 🙂

File: src/Chapter2.hs (hunk 3 above)
Comment (vrom911) [reply src/Chapter2.hs]: You can implement it using `foldr`, and you will need to use `++` anyway. So these two solutions are based on the base approach 🙂

File: src/Chapter2.hs (hunk 6 above)
Comment (ynfle) [reply src/Chapter2.hs]: I improved using a local function. Let me know what you think

File: src/Chapter2.hs (hunk 4 above)
Comment (ynfle) [reply src/Chapter2.hs]: Done

File: src/Chapter2.hs (hunk 3 above)
Comment (ynfle) [reply src/Chapter2.hs]: Done

**Comment by vrom911** [comment]
@ynfle Wonderful!
This is ready for merge 🚢

**Pull request merged**